pkg/egressgateway/manager.go

```diff
@@ -191,104 +152,93 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
-	}
 	delete(manager.epDataStore, id)
-}
 
-func getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint) (*endpointMetadata, error) {
-	var ipv4s []string
-	id := types.NamespacedName{
-		Name:      endpoint.GetName(),
-		Namespace: endpoint.GetNamespace(),
-	}
+	manager.reconcile()
+}
 
-	if endpoint.Networking == nil {
-		return nil, fmt.Errorf("endpoint has no networking metadata")
-	}
+func (manager *Manager) addMissingEgressRules() {
+	egressPolicies := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
+	egressmap.EgressPolicyMap.IterateWithCallback(
+		func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
+			egressPolicies[*key] = *val
+		})
+
+	for _, policyConfig := range manager.policyConfigs {
+		for _, endpoint := range manager.epDataStore {
+			if !policyConfig.selectsEndpoint(endpoint) {
+				continue
+			}
 
-	for _, pair := range endpoint.Networking.Addressing {
-		if pair.IPV4 != "" {
-			ipv4s = append(ipv4s, pair.IPV4)
+			for _, endpointIP := range endpoint.ips {
+				for _, dstCIDR := range policyConfig.dstCIDRs {
+					policyKey := egressmap.NewEgressPolicyKey4(endpointIP, dstCIDR.IP, dstCIDR.Mask)
+					policyVal, policyPresent := egressPolicies[policyKey]
+
+					if policyPresent && policyVal.Match(policyConfig.egressIP, policyConfig.egressIP) {
+						continue
+					}
+
+					err := egressmap.ApplyEgressPolicy(endpointIP, *dstCIDR, policyConfig.egressIP, policyConfig.egressIP)
+					if err != nil {
+						log.WithError(err).WithFields(logrus.Fields{
+							logfields.SourceIP:        endpointIP,
+							logfields.DestinationCIDR: *dstCIDR,
+							logfields.EgressIP:        policyConfig.egressIP,
+							logfields.GatewayIP:       policyConfig.egressIP,
+						}).Error("Error applying egress policy.")
+					}
+				}
+			}
 		}
 	}
-
-	if endpoint.Identity == nil {
-		return nil, fmt.Errorf("endpoint has no identity metadata")
-	}
-
-	data := &endpointMetadata{
-		ips:    ipv4s,
-		labels: labels.NewLabelsFromModel(endpoint.Identity.Labels).K8sStringMap(),
-		id:     id,
-	}
-
-	return data, nil
 }
 
-// upsertPolicyEndpoint updates or insert to endpoint policy mapping for given policy config and endpoints,
-// it also upserts egress map to keep in sync
-func (manager *Manager) upsertPolicyEndpoint(config *PolicyConfig, epData *endpointMetadata) error {
-	if err := manager.updateEgressMap(epData.ips, config); err != nil {
-		return err
-	}
+func (manager *Manager) removeUnusedEgressRules() {
+	egressPolicies := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
+	egressmap.EgressPolicyMap.IterateWithCallback(
+		func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
+			egressPolicies[*key] = *val
+		})
+
+nextPolicyKey:
+	for policyKey, policyVal := range egressPolicies {
+		for _, policyConfig := range manager.policyConfigs {
+			for _, endpoint := range manager.epDataStore {
+				if !policyConfig.selectsEndpoint(endpoint) {
+					continue
+				}
```

Comment:
Nit: These 5 lines could be moved outside the `egressPolicies` loop, to avoid computing the same thing for each egress policy.

Reply:
I don't think it's possible as for each entry in the bpf map we necessarily need to test all k8s policies. The pseudocode would be:

```
nextEntry:
for each entry in the egress_policy map {
    for each policy in k8s egressnatpolicies {
        if policy matches entry {
            // we found one k8s policy that matches the current BPF entry, move to the next one
            continue nextEntry
        }
    }

    // the current BPF entry is not backed by any k8s policy, delete it
    egressmap.RemoveEgressPolicy()
}
```

(while if we start looping from k8s policies, we would not be able to tell if a given BPF entry needs to be deleted or not)

I'll add a comment to clarify this